src/rust/engine/process_execution/src/cache.rs

```diff
@@ -114,20 +115,40 @@ impl CommandRunner {
       })
       .await?;
 
-    if let Some((execute_response, platform)) = maybe_execute_response {
+    // Deserialize the cache entry if it existed.
+    let result = if let Some((execute_response, platform)) = maybe_execute_response {
       crate::remote::populate_fallible_execution_result(
         self.file_store.clone(),
         execute_response.get_result(),
         vec![],
         platform,
         true,
       )
-      .map(Some)
       .compat()
-      .await
+      .await?
     } else {
-      Ok(None)
-    }
+      return Ok(None);
+    };
+
+    // Ensure that all digests in the result are loadable, erroring if any are not.
+    let _ = future03::try_join_all(vec![
+      self
+        .file_store
+        .ensure_local_has_file(result.stdout_digest)
+        .boxed(),
+      self
+        .file_store
+        .ensure_local_has_file(result.stderr_digest)
+        .boxed(),
+      self
+        .file_store
+        .ensure_local_has_recursive_directory(result.output_directory)
```

Comment:
This is actually a much stronger guarantee than I think we _need_, and may actually make cache hits slower than builds in the case of fast-ish actions with large-ish outputs... There's a potential optimisation here to merge the output of `store::remote::Store::list_missing_digests` and the currently not-existing `store::local::Store::list_missing_digests` to test for missing-ness, rather than forcing an entire download.

Reply:
Yea, agree that this could be improved... the downloading sideeffect is not actually desirable here, but the method existed. Another factoring I thought of after landing was:
```
let local_only_store = self.file_store.local_only();
local_only_store.ensure_local_has_recursive_directory(..)
```

Are you saying that `ensure_local_has_recursive_directory` should be reimplemented in terms of comparing `list_missing_digests` methods? That sounds reasonable. You'd still need to do a recursive walk first to find all relevant digests before running that method, but it would avoid "loading" in favor of the equivalent of a "batch exists" method. Or we could switch to storing/using the `Tree` object to remove all of the round trips...

Reply:
> Are you saying that `ensure_local_has_recursive_directory` should be reimplemented in terms of comparing `list_missing_digests` methods? That sounds reasonable. You'd still need to do a recursive walk first to find all relevant digests before running that method, but it would avoid "loading" in favor of the equivalent of a "batch exists" method.

Yeah (tbh I thought it was already implemented that way, but apparently not!) - we can still start kicking off a few blob fetches while we do the Directory walk if we find we're bottlenecked on it. Then this method would only use the first half of that, without doing the follow-up fetches.